docs/dev/hacking-guide.md

```diff
@@ -0,0 +1,233 @@
+```
+apiVersion: v1
+kind: ConfigMap
+data:
+  images.json: |
+    {
+        "machineAPIOperator": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerAWS": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerOpenStack": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerLibvirt": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerBareMetal": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerAzure": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerGCP": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerOvirt": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "clusterAPIControllerVSphere": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalOperator": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalIronic": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalIronicInspector": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalIpaDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalMachineOsDownloader": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+        "baremetalStaticIpManager": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:..."
+    }
+```
+
+```
+oc edit configmap/machine-api-operator-images -n openshift-machine-api
+```
+
+With the new image information loaded into the ConfigMap, the next thing you might do is replace the Machine API operator. This operator controls how the specific cloud controllers are deployed and coordinated. You only change this component if there is something you are testing.
+
+The easiest way to change this operator is to scale down the deployment, change the image reference in the deployment, and then scale it back up. The commands you can use:
+
+```
+oc scale --replicas=0 deploy/machine-api-operator -n openshift-machine-api
+oc edit deploy/machine-api-operator -n openshift-machine-api
+oc scale --replicas=1 deploy/machine-api-operator -n openshift-machine-api
+```
+
+After changing the download reference in the images ConfigMap an easy way to swap out the controller is to let the Machine API operator do it for you. You can delete the deployment associated with the cloud provider controller and then the Machine API operator will create a new one for you, like this:
+
+```
+oc delete deploy/machine-api-controllers -n openshift-machine-api
+```
+
+## How to run e2e tests
+Our e2e test suite is located in this repository https://github.com/openshift/cluster-api-actuator-pkg. 
```

Comment:
i think we should probably point to the readme in the cluster-api-actuator-pkg repo for instructions on running those tests. it will get cumbersome if we need to update both places.